core/domain/expression_parser.py

```diff
@@ -0,0 +1,541 @@
+            if operator_token.text == '+':
+                parsed_expr = AddOperator(parsed_expr, parsed_right)
+            else:
+                parsed_expr = SubtractOperator(parsed_expr, parsed_right)
+            operator_token = self.is_next_token(['+', '-'])
+        return parsed_expr
+
+    def parse_mul_expr(self):
+        """Function representing the following production rule of the grammar:
+        <mul_expr> ::= <pow_expr> (('*' | '/') <pow_expr>)*
+
+        Returns:
+            Node. Root node of the generated parse tree.
+        """
+
+        parsed_expr = self.parse_pow_expr()
+        operator_token = self.is_next_token(['*', '/'])
+        while operator_token:
+            parsed_right = self.parse_pow_expr()
+            if operator_token.text == '*':
+                parsed_expr = MultiplyOperator(parsed_expr, parsed_right)
+            else:
+                parsed_expr = DivideOperator(parsed_expr, parsed_right)
+            operator_token = self.is_next_token(['*', '/'])
+        return parsed_expr
+
+    def parse_pow_expr(self):
+        """Function representing the following production rule of the grammar:
+        <pow_expr> ::= '-' <pow_expr> | '+' <pow_expr> |
+        <unit> ('^' <pow_expr>)?
+
+        Returns:
+            Node. Root node of the generated parse tree.
+        """
+
+        # Eliminate any leading unary '+' or '-' operators, because they
+        # are syntactically irrelevant.
+        while self.is_next_token(['+', '-']):
+            pass
+
+        parsed_expr = self.parse_unit()
+        operator_token = self.is_next_token(['^'])
+        if operator_token:
+            # Using recursion for right-associative ^ operator.
+            parsed_right = self.parse_pow_expr()
+            return PowerOperator(parsed_expr, parsed_right)
+        return parsed_expr
+
+    def parse_unit(self):
+        """Function representing the following production rule of the grammar:
+        <unit> ::= <identifier> | <number> | '(' <expr> ')' |
+        <function> '(' <expr> ')'
+
+        Returns:
+            Node. Root node of the generated parse tree.
+
+        Raises:
+            Exception: Invalid token.
+        """
+
+        token = self.get_next_token()
+        if token.category == 'identifier':
+            return Identifier(token)
+
+        if token.category == 'function':
+            if self.is_next_token(['(']):
+                parsed_child = self._parse_expr()
+                self.expect_token(')')
+                return Function(token, parsed_child)
+
+        if token.category == 'number':
+            return Number(token)
+
+        if token.text == '(':
+            parsed_expr = self._parse_expr()
+            self.expect_token(')')
+            return parsed_expr
+
+        raise Exception('Invalid token: %s.' % token.text)
+
+    def check_next_token(self):
+        """Function to peek into the next position to see the next token.
```

Comment:
Add that it does without iterating `next_token_index`, since that seems to be the main part?

Reply:
Done.